job-runner/chain/reaper.go

```diff
@@ -0,0 +1,630 @@
+		},
+		runJobChan: f.RunJobChan,
+	}
+}
+
+// Make a JobReaper for use on a job chain being suspended.
+func (f *ChainReaperFactory) MakeSuspended() JobReaper {
+	return &SuspendedChainReaper{
+		reaper: reaper{
+			chain:             f.Chain,
+			chainRepo:         f.ChainRepo,
+			rmc:               f.RMClient,
+			logger:            f.Logger,
+			finalizeTries:     f.RMCTries,
+			finalizeRetryWait: f.RMCRetryWait,
+			doneJobChan:       f.DoneJobChan,
+			stopChan:          make(chan struct{}),
+			doneChan:          make(chan struct{}),
+			stopMux:           &sync.Mutex{},
+		},
+		runnerRepo: f.RunnerRepo,
+	}
+}
+
+// Make a JobReaper for use on a job chain being stopped.
+func (f *ChainReaperFactory) MakeStopped() JobReaper {
+	return &StoppedChainReaper{
+		reaper: reaper{
+			chain:             f.Chain,
+			chainRepo:         f.ChainRepo,
+			rmc:               f.RMClient,
+			logger:            f.Logger,
+			finalizeTries:     f.RMCTries,
+			finalizeRetryWait: f.RMCRetryWait,
+			doneJobChan:       f.DoneJobChan,
+			stopChan:          make(chan struct{}),
+			doneChan:          make(chan struct{}),
+			stopMux:           &sync.Mutex{},
+		},
+		runnerRepo: f.RunnerRepo,
+	}
+}
+
+// -------------------------------------------------------------------------- //
+
+// Job Reaper for running chains.
+type RunningChainReaper struct {
+	reaper
+	runJobChan chan proto.Job // enqueue next jobs to run here
+}
+
+// Run reaps jobs when they finish running. For each job reaped, if...
+// - chain is done: save final state + send to RM.
+// - job failed:    retry sequence if possible.
+// - job completed: prepared subsequent jobs and enqueue if runnable.
+func (r *RunningChainReaper) Run() {
+	defer close(r.doneChan)
+
+	// If the chain is already done, skip straight to finalizing.
+	done, complete := r.chain.IsDoneRunning()
+	if done {
+		r.Finalize(complete)
+		return
+	}
+
+REAPER:
+	for {
+		select {
+		case job := <-r.doneJobChan:
+			r.Reap(job)
+			done, complete = r.chain.IsDoneRunning()
+			if done {
+				break REAPER
+			}
+		case <-r.stopChan:
+			// Don't Finalize the chain when stopping - the stopped or suspended
+			// reaper will take care of that.
+			return
+		}
+	}
+
+	r.Finalize(complete)
+}
+
+// Stop stops the reaper from reaping any more jobs. It blocks until the reaper
+// is stopped (will reap no more jobs and Run will return).
+func (r *RunningChainReaper) Stop() {
+	r.stopMux.Lock()
+	defer r.stopMux.Unlock()
+	if r.stopped {
+		return
+	}
+	r.stopped = true
+
+	close(r.stopChan)
+	<-r.doneChan
+	return
+}
+
+// reap takes a job that just finished running, saves its final state, and prepares
+// to continue running the chain (or recognizes that the chain is done running).
+//
+// If chain is done: save final state + stop running more jobs.
+// If job failed:    retry sequence if possible.
+// If job completed: prepared subsequent jobs and enqueue if runnable.
+func (r *RunningChainReaper) Reap(job proto.Job) {
```

Comment:
Does this method need to be public? I don't see it defined on the interface. Or is this just to make testing easier?

Reply:
It's just for testing.